### PR TITLE
Exercise_2

### PR DESCRIPTION
**Pull request by kushal-7**
Acho, this is from exercise 2
There are edit feature with form validation, in memory server, delete feature and search functionality

### CODE REVIEW

File: src/app/dashboard/dashboard.component.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  heroes: Hero[] = [];
+
+  constructor(private heroService: HeroService) { }
+
+  ngOnInit(): void {
+    this.getHeroes();
+  }
+
+  getHeroes(): void {
+    this.heroService.getHeroes()
```
Comment (ThinleyJimmyDorji) [inline src/app/dashboard/dashboard.component.ts:20]: cool

File: src/app/hero-form/hero-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+export class HeroFormComponent {
+
+  powers = ['Really Smart', 'Super Flexible',
+    'Super Hot', 'Weather Changer'];
+
+  model = new Hero(18, 'Dr. IQ', this.powers[0], 'Chuck Overstreet');
+
+  submitted = false;
+
+  onSubmit() { this.submitted = true; }
```
Comment (ThinleyJimmyDorji) [inline src/app/hero-form/hero-form.component.ts:19]: It's always good to provide a return type. Standard practice. Install typescript for that.
```javascript
onSubmit(): void {
//blah blah
}
```

File: src/app/hero-search/hero-search.component.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import { HeroService } from '../hero.service';
+
+@Component({
+  selector: 'app-hero-search',
+  templateUrl: './hero-search.component.html',
+  styleUrls: ['./hero-search.component.css']
+})
+export class HeroSearchComponent implements OnInit {
+  heroes$!: Observable<Hero[]>;
+  private searchTerms = new Subject<string>();
```
Comment (ThinleyJimmyDorji) [inline src/app/hero-search/hero-search.component.ts:19]: cool.

File: src/app/hero-search/hero-search.component.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  ngOnInit(): void {
+    this.heroes$ = this.searchTerms.pipe(
+      // wait 300ms after each keystroke before considering the term
+      debounceTime(300),
+
+      // ignore new term if same as previous term
+      distinctUntilChanged(),
+
+      // switch to new search observable each time the term changes
+      switchMap((term: string) => this.heroService.searchHeroes(term)),
```
Comment (ThinleyJimmyDorji) [inline src/app/hero-search/hero-search.component.ts:37]: nice

File: src/app/hero.service.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      tap(_ => this.log(`fetched hero id=${id}`)),
+      catchError(this.handleError<Hero>(`getHero id=${id}`))
+    );
+  }
+
+  /* GET heroes whose name contains search term */
+  searchHeroes(term: string): Observable<Hero[]> {
+    if (!term.trim()) {
+      // if not search term, return empty hero array.
+      return of([]);
```
Comment (ThinleyJimmyDorji) [inline src/app/hero.service.ts:60]: cool

File: src/app/hero.service.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    return this.http.get<Hero[]>(`${this.heroesUrl}/?name=${term}`).pipe(
+      tap(x => x.length ?
+        this.log(`found heroes matching "${term}"`) :
+        this.log(`no heroes matching "${term}"`)),
+      catchError(this.handleError<Hero[]>('searchHeroes', []))
+    );
+  }
```
Comment (ThinleyJimmyDorji) [inline src/app/hero.service.ts:68]: it's good to leave only a line space between class declarations. improves readability

File: src/app/heroes/heroes.component.ts [hunk 7]
```diff
@@ -1,40 +1,39 @@
...
+      .subscribe(hero => {
+        this.heroes.push(hero);
+      });
   }
-  
-}
 
+  delete(hero: Hero): void {
+    this.heroes = this.heroes.filter(h => h !== hero);
+    this.heroService.deleteHero(hero.id).subscribe();
```
Comment (ThinleyJimmyDorji) [inline src/app/heroes/heroes.component.ts:36]: use `switchMap` operator to fetch new collection of heroes after a delete operation. It makes your feature reactive.

File: src/app/hero.service.ts (hunk 6 above)
Comment (kushal-7) [reply src/app/hero.service.ts:68]: sure acho I will do that

**Pull request merged**